Deferrable mode for EksCreateFargateProfileOperator and EksDeleteFargateProfileOperator

syedahsn opened the pull request:
This PR allows deferrable mode to be used with the `EksCreateFargateProfileOperator` and the `EksDeleteFargateProfileOperator`. It also includes unit tests to test the triggers for both of the operators.
<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

syedahsn committed:
Deferrable mode for EksCreateFargateProfileOperator and EksDeleteFargateProfileOperator
Unit tests for both

vandonr-amz commented on airflow/providers/amazon/aws/operators/eks.py:
```diff
@@ -401,13 +417,31 @@ def execute(self, context: Context):
             self.log.info("Waiting for Fargate profile to provision.  This will take some time.")
             eks_hook.conn.get_waiter("fargate_profile_active").wait(
                 clusterName=self.cluster_name, fargateProfileName=self.fargate_profile_name
             )
```
shall we pass the delay & max attempts to this waiter as well while we're at it ?

ferruzzi replied:
+1, I would say that is implied.  Not having been part of this project, my expectation was that those settings would be applied here as well, seems reasonable that a user would as well.

vandonr-amz commented on airflow/providers/amazon/aws/operators/eks.py:
```diff
@@ -401,13 +417,31 @@ def execute(self, context: Context):
+                    max_attempts=self.waiter_max_attempts,
+                ),
+                method_name="execute_complete",
+                timeout=timedelta(seconds=(self.waiter_max_attempts * self.waiter_delay + 60)),
```
why 60 ? Maybe create a named constant ?

syedahsn replied:
I don't know if we need a named constant for this. 60 is arbitrary - I think adding a new constant will just add another variable someone will need to look up the value of.

vandonr-amz replied:
how about a comment then ?
```suggestion
                # adding 60s to theoretical value to have extra room to account for time lost here and there
                timeout=timedelta(seconds=(self.waiter_max_attempts * self.waiter_delay + 60)),
```

syedahsn replied:
I have a comment that I used in the redshift operators. I'll use that to be consistent:

```
# timeout is set to ensure that if a trigger dies, the timeout does not restart
# 60 seconds is added to allow the trigger to exit gracefully (i.e. yield TriggerEvent)
```

vandonr-amz commented on airflow/providers/amazon/aws/operators/eks.py:
```diff
@@ -622,12 +667,31 @@ def execute(self, context: Context):
         eks_hook.delete_fargate_profile(
             clusterName=self.cluster_name, fargateProfileName=self.fargate_profile_name
         )
-        if self.wait_for_completion:
```
same comments as above to this code

vandonr-amz commented on airflow/providers/amazon/aws/triggers/eks.py:
```diff
@@ -0,0 +1,178 @@
+
+    @cached_property
+    def hook(self) -> EksHook:
+        return EksHook(aws_conn_id=self.aws_conn_id)
```
mentioned it earlier, but I think caching this is useless

vandonr-amz commented on airflow/providers/amazon/aws/triggers/eks.py:
```diff
@@ -0,0 +1,178 @@
+                    if "terminal failure" in str(error):
+                        yield TriggerEvent(
+                            {"status": "failure", "message": f"Create Fargate Profile failed: {error}"}
+                        )
```
you should raise directly here instead of returning the error to the operator
see https://apache-airflow.slack.com/archives/CCPRP7943/p1685663968275419

vandonr-amz commented on airflow/providers/amazon/aws/triggers/eks.py:
```diff
@@ -0,0 +1,178 @@
+            yield TriggerEvent({"status": "success", "message": "Fargate Profile Created"})
+
+
+class EksDeleteFargateProfileTrigger(BaseTrigger):
```
same comments as above

ferruzzi commented on airflow/providers/amazon/aws/operators/eks.py:
```diff
@@ -371,11 +381,14 @@ def __init__(
         region: str | None = None,
+        waiter_delay: int = 60,
+        waiter_max_attempts: int = 30,
+        deferrable: bool = False,
```
Not for this PR, but another argument for a BaseAwsOperator would be to offload a bunch of these to a central place instead of repeating them constantly in every operator.  The last 6 are fairly universal.

syedahsn replied:
There's a [PR](https://github.com/apache/airflow/pull/31712#pullrequestreview-1476832822) open to have the `deferrable` parameter being inherited from `BaseDeferrableOperator`. Nothing for the other operators yet

ferruzzi commented on airflow/providers/amazon/aws/operators/eks.py:
```diff
@@ -371,11 +381,14 @@ def __init__(
         aws_conn_id: str = DEFAULT_CONN_ID,
         region: str | None = None,
+        waiter_delay: int = 60,
+        waiter_max_attempts: int = 30,
```
Are these values arbitrary?  If they are, maybe have a look at the botocore code and see what the default official waiter values are.

syedahsn replied:
I had the defaults backwards. I'll change that. Thanks!

ferruzzi commented on airflow/providers/amazon/aws/triggers/eks.py:
```diff
@@ -0,0 +1,178 @@
+
+    def serialize(self) -> tuple[str, dict[str, Any]]:
+        return (
+            self.__class__.__module__ + "." + self.__class__.__qualname__,
```
Again, not necessarily for this PR:  Looks like some Triggers use this and others have a fixed string. Might be a case for making a triggers/utils.py and putting in a get_trigger_name() helper?

syedahsn replied:
This is an improvement to what we were initially doing, which was a hard coded string with the trigger path. As for making it a separate function, I'm open to it if others feel it necessary, but considering its a one line function, I don't feel like its worth it. But I'm not opposed to it either.

ferruzzi commented on tests/providers/amazon/aws/triggers/test_eks.py:
```diff
@@ -0,0 +1,312 @@
+    async def test_eks_create_fargate_profile_trigger_run_multiple_attempts(
+        self, mock_async_conn, mock_sleep
+    ):
+        a_mock = mock.MagicMock()
```
It seems like a lot of these tests are reusing some setup code that could go into a `setup_method`

syedahsn replied:
Thats a good point. My only concern is that often there are minor changes between each setup. I'll see if there's a way to package it all in a setup method

ferruzzi reviewed (commented):
Left a few non-binding suggestions.

syedahsn committed:
Remove hook caching in Trigger
Raise Error from Trigger rather than a TriggerEvent in case of failure
Other minor fixes

syedahsn committed:
Add EKS trigger to provider.yaml

syedahsn committed:
Add comment on operator timeout

vandonr-amz reviewed (approved):
lgtm

syedahsn committed:
Fix integration name for EKS for triggers in provider.yaml

o-nikolas reviewed (approved):
Browsed the code quickly, and there are some other reviews/approvals. Merging this one, we can always circle back later
